src/NUnitEngine/nunit.engine.tests/Agents/RemoteTestAgentTests.cs

```diff
@@ -0,0 +1,93 @@
+﻿// ***********************************************************************
+// Copyright (c) 2016 Charlie Poole
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+// 
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using NUnit.Framework;
+
+namespace NUnit.Engine.Agents.Tests
+{
+    [TestFixture]
+    public sealed class RemoteTestAgentTests
+    {
+        [Test]
+        public void Stop_ShouldUnregisterWithAgency()
+        {
+            List<ITestAgent> unregisteredAgents = new List<ITestAgent>();
+
+            ITestAgency agency = new MockAgency(
+                register: agent => { Assert.Fail("Should not be invoked"); },
+                unregister: unregisteredAgents.Add
+            );
+
+            using (RemoteTestAgent testAgent = new RemoteTestAgent(Guid.NewGuid(), agency, services: null))
+            {
+                testAgent.Stop();
+
+                Assert.AreEqual(1, unregisteredAgents.Count, "Agent should only be unregistered once");
+                Assert.AreSame(testAgent, unregisteredAgents[0], "Agent should pass itself to agency");
+            }
+        }
+
+        [Test]
+        public void Stop_WhenAgencyThrows_ShouldNotThrow()
+        {
+            ITestAgency agency = new MockAgency(
+                register: agent => { Assert.Fail("Should not be invoked"); },
+                unregister: agent => { throw new Exception("Boom"); }
+            );
```

Comment:
Another case where this is super useful is when you're passing things like null or 0 to a method.  If you look below, had I constructed RemoteTestAgent with null for the services, you put the work on the reader to figure out what null means.  But using the named parameter it's clearer, and it calls out my intent.